### ISSUE TITLE
introduce torch.Scalar to represent scalars in autograd

### ISSUE THREAD

**Issue by soumith** [pytorch/pytorch#1433]
Currently, scalars are returned in autograd as 1-element Tensors of dim=1. 
This has various unintended side effects and issues.
We need to introduce a scalar type into either torch or autograd to avoid the side-effects.

Some of the known side-effects:
- indexing is infinitely chained: a[0][0][0][0][...]
- overflows are not handled as expected ( https://github.com/pytorch/pytorch/issues/1389 )

**Comment by oleg-trott** [comment]
Scalars are tensors of dim=0.

PyTorch (unlike Numpy) seems to have an internally inconsistent interpretation here:

```
>>> shape = [2, 3]
>>> torch.ones(shape).nelement() == np.prod(shape)
True

>>> shape = []
>>> torch.ones(shape).nelement() == np.prod(shape)
False
```

If this (as well as broadcasting) were fixed, perhaps there wouldn't be a need for the special `torch.Scalar` class?

**Comment by gchanan** [comment]
Numpy has both "Scalars" and 0-dimensional arrays, although Scalar may just be a wrapper over a 0-dimensional array.

```
>>> x=np.sum(np.random.randn(1),0)
>>> type(x)
numpy.float64
>>> x.shape
()
>>> y=np.squeeze(np.random.randn(1), 0)
>>> type(y)
numpy.ndarray
>>> y.shape
()
```

I haven't looked into this closely, it may just be a convenience type, i.e. users expects sum() to return a Scalar but squeeze to return a Tensor.

**Comment by zym1010** [comment]
@gchanan based on <https://stackoverflow.com/questions/773030/why-are-0d-arrays-in-numpy-not-considered-scalar>, I think numpy has both 0-d array and scalar, just for interoperability with other Python libraries, such as `math`. If numpy can be designed without consideration of other libraries, I bet only 0-d array will be there.

I think it's easier to simply only have 0-d array, as done in TensorFlow.

**Comment by gchanan** [comment]
I don't think a Scalar type is really the correct solution to the overflow problem (https://github.com/pytorch/pytorch/issues/1389); that would fix the all dimension case, but not the case where a single dimension is specified.  The right solution, I think, is to have a dtype parameter that specifies the output type (and maybe accumulation type as well, although we are limited by TH/THC here).

**Comment by gchanan** [comment]
I had a discussion with @colesbury, @apaszke, @ezyang about this.  Here are some notes from that discussion followed by my own proposal.  The below discussion is limited to “Scalars” in autograd; we probably want to move the Tensor and Variable API together in some some sensible way (see https://github.com/pytorch/pytorch/issues/2633), but I don’t want to muddy the discussion of autograd Scalars with the details of that process as much as is feasible.

**Scalars vs 0-dimensional objects**

NumPy has both scalars and 0-dimensional arrays.  Scalars are immutable numeric types that (at least for cases where the type is compatible) are instances of the relevant python numeric type, e.g.:
```
>>> isinstance(np.double(5), float)
True
>>> isinstance(np.float(5),float)
True
```
They also “quack” like 0-dimensional arrays, i.e. they have `shape` and `ndim` attributes:
```
>>> np.double(5).shape
()
>>> np.double(5).ndim
0
```
These properties allow you to pass scalar types to libraries outside of NumPy (even ones that do explicit type checking via isinstance or e.g. PyFloat_Check) and to mix NumPy scalars and nd-arrays in code (as long as you don’t try to mutate a scalar).

On the other handle, 0-dimensional arrays are mutable types that are not instances of python numeric types:
```
>>> a=np.array(0.5)
>>> a[()]=2
>>> a
array(2.0)
>>> isinstance(np.array(0.5), float)
False
>>> isinstance(np.array(0.5), np.ndarray)
True
```
Thus, at least for libraries that do explicit type checking, one would need to cast the 0-dimensional array to the correct python or NumPy scalar type before converting it (np.asscalar will do this)

**Do we want Scalars or 0-dimensional Variables or both in PyTorch?**

I’m going to argue that we only want 0-dimensional Variables.

First, we want at least 0-dimensional Variables in PyTorch.  Consider if we had only (immutable) Scalars and your loss is a .sum() over some variable.  Now, if you treat this as an immutable scalar, either it has a grad and can be optimized (which is weird for an immutable value), or it doesn’t (and you have to hack around that in some way to still be able to do training).  You could also represent it as a 1-dimensional Variable (which is what we do today), but then the dimensions are inconsistent (i.e. reduction functions reduce the dimension by 1 except when the dimension is 1).  Representing this quantity as a 0-dimensional Variable is just much more straightforward and mathematically consistent.

Now, if we have 0-dimensional Variables, do we need immutable scalars in autograd?  I’d argue no:
1) The ability to use Scalars in external libraries seem like not a strong reason for the extra complexity: many operations should just work on 0-dimensional arrays (because we define a number of numerical operators on Variables) and if we provide conversions, in the worst case users can just call the conversions.
2) Immutability in-and-of-itself may be useful, but its usefulness is not just restricted to scalar types; you may want immutability on all tensor/Variable types.
3) A related question is what to return when indexing the last dimension of a Variable.  The normal python, numpy, and current Tensor semantics are that it gives you an (immutable) scalar type, while the current Variable semantics are that it gives you a Variable that shares storage (of the wrong dimensionality).  Given that we already return Variables in this case, returning a 0-dimensional one seems acceptable.

I’ll also note here that TensorFlow only supports 0-dimensional tensors, although it’s quite different because the execution doesn’t define the graph, as in PyTorch.

Many of the other arguments for scalars are not about immutability, but convenience, e.g. it’s nice to get pretty prints (i.e. “0” or “torch.FloatScalar(0)” instead of something like:
 ```
 5
[torch.FloatTensor of size ()]
```

Those kinds of convenience problems are solvable without adding scalar types.

**A Proposal for adding 0-dimensional support to autograd**


As mentioned above, the goal here is to get to a sensible Variable API; this doesn’t purport to solve all issues with scalars (i.e. forced synchronization), but that we can work on those issues separately while maintaining the API as is.

1) Autograd forward functions can now return a python number to represent that a 0-dimensional Variable should be returned to the .apply caller.  Note that they can’t return 0-dimensional objects directly because the forward functions are written in terms of tensors, which don’t support 0-dimensions yet.  Note also that this will now match what the equivalent tensor functions do (i.e. `max()` returns a python number), and when they support 0-dimensional tensors, the code will just work as written.  This won’t immediately solve all dimensionality problems, because for example `max(dim=0)` on a 1-dimensional tensor will still return a 1-dimensional tensor, but again, this will automatically be fixed when the tensor APIs are changed.

2) Variable is now 0-dimensional aware (via its underlying ATen implementation).  For example:
`.dim()` will return 0
`.size()` will return ()
`.shape` will return ()
indexing with a dimension will throw an error (i.e. `[()]` and `[...]` will work, everything else will throw an error)
All other functions will behave as before.  Note that accessing .data will return a 1-dimensional tensor, but again, that will automatically change when tensor changes.

3) We will provide constructors for 0-dimensional Variables.  Unfortunately, a python number passed to a tensor constructor is interpreted as a size: i.e. `torch.FloatTensor(5)` creates a 1-dimensional tensor of size 5, so we eventually need a different name when we support 0-dimensions on tensors.  I propose torch.*Scalar (e.g. `torch.FloatScalar`); this has a more object-looking name than say, torch.float(...), which will make it more obvious it’s a mutable type.  One question is how this should interact with the work in (https://github.com/pytorch/pytorch/issues/2633).  A full “leap-of-faith” approach would be to have this “constructor” always return a Variable and have ‘requires_grad’ and ‘volatile’ parameters.  This name will probably be confusing in the short term because you would expect `torch.FloatScalar` to return a tensor type, but constructing these objects directly will probably be rare in the short term anyway.

4) `__str__` on a 0-dimensional tensor will return the underlying number, `__repr__` will return how it is constructed, i.e. `torch.FloatScalar(5)`.  This is consistent with numpy 0-dimensional arrrays:
```
>>> print(np.array(0))
0
>>> np.array(0)
array(0)
```

5) We will provide conversion functions (`__int__`, `__float__`, more?) for 0-dimensional Variables for ease of use with existing python libraries.  Calling these functions on non-0-dimensional Variables will throw an error.

Thoughts?

**Comment by gchanan** [comment]
After some further discussion, I'd propose the following change:

Instead of introducing torch.*Scalar, we introduce a data-only factory like `np.array(...)` so that the specification of a python number is unambiguous: it represents the data, so creates a Scalar.  For convenience, this factory will also take a `dtype` (and `device`).  Let's call this `torch.tensor` (I'm open to different names) Examples:

```
 # create a 0-dim tensor of default type with data "5"
torch.tensor(5)

# create a 0-dim tensor of type torch.DoubleTensor with data "1.5"
torch.tensor(1.5, dtype=torch.double)

 #create a 2x2 torch.cuda.LongTensor on device 1 with data from [0,3] increasing.
torch.tensor([[0, 1], [2,3]], dtype=torch.cuda.long, device=1)
```

**Comment by zym1010** [comment]
@gchanan so for scalar vs. 0-d array, do you think that numpy people have scalar just for the sake of compatibility of Python builtin functions and stdlib?

**Comment by fmassa** [comment]
@gchanan that was exactly what I was going to propose, but I was afraid it would be a bit too much work if we wanted to keep things similar to numpy.
In general, there has been a [lot of demand for a numpy-like api](https://github.com/pytorch/pytorch/issues/2228), and passing a `dtype` is definitely something we would like to have. Now, do we want it to be `tensor` or `array`?

**Comment by gchanan** [comment]
@zym1010 that's what the stackoverflow link you provided implies.  I read through some old numpy discussion threads and it's still a bit unclear to me; talks a lot about legacy from `Numeric` that isn't really relevant to us.  One thing I like about only providing 0-dim tensors is that we can add (immutable) scalars later if there is a compelling reason as long as they are duck-typed correctly and we avoid excessive type checking.  If we jump to providing (immutable) scalars and they aren't useful we have to support both.

**Comment by gchanan** [comment]
Glad to see we are thinking along the same general lines @fmassa.  On the amount of work question, I'm really only proposing that we define a `dtype` that can be passed to non-typed torch constructors/factory, e.g. `randn`, `ones`, `zeros`, and the new `tensor` (?), and maybe changing the string representation to print a `dtype` so that (at least small) tensors can be constructed directly from that representation.

I'm not really sure I want to weigh in on the `array` vs `tensor` naming, it's a good question, but as you noted, would be a lot of work to change.

**Comment by vadimkantorov** [comment]
For naming, my two cents is that it's better to avoid collision between `torch.tensor` and already existing `torch.Tensor`, especially if their interface would be mostly similar.

**Comment by obust** [comment]
I think it would make sense to have `torch.Tensor` as the tensor factory.
Today it is an alias to the `torch.FloatTensor` but it doesn't have to be.
It can very well be :

```
Tensor(1) # 0-dimensional FloatTensor
Tensor([1, 1]) # 1-dimensional FloatTensor
Tensor([1, 1], dtype=torch.double) # 1-dimensional DoubleTensor
```
This way the instanciation API remains class-based, and you can specialize the data type with a `dtype` argument. I am pretty sure this is possible using python metaclass. https://stackoverflow.com/a/23377870/4577665

**Comment by soumith** [comment]
scalars are merged into master by @gchanan . This can be closed.
